python/lsst/obs/base/ingest.py

```diff
@@ -318,16 +365,181 @@ def _calculate_dataset_info(self, header, filename):
+                                 f" {missing} (via JSON)")
+
+        else:
+            obsInfo = ObservationInfo(header, pedantic=False, filename=filename,
+                                      required={k for k in ingest_subset if ingest_subset[k]},
+                                      subset=set(ingest_subset))
 
         dataId = DataCoordinate.standardize(instrument=obsInfo.instrument,
                                             exposure=obsInfo.exposure_id,
                                             detector=obsInfo.detector_num,
                                             universe=self.universe)
         return RawFileDatasetInfo(obsInfo=obsInfo, dataId=dataId)
 
+    def expandIndexFiles(self, files):
+        """Given a list of files, look for index files and read them.
+
+        Parameters
+        ----------
+        files : iterable over `str` or path-like objects
+            Paths to the files to be ingested.  Will be made absolute
+            if they are not already.
+
+        Returns
+        -------
+        index : `dict[str, Any]`
+            Merged contents of all relevant index files found. These can
+            be explicitly specified index files or ones found in the
+            directory alongside a data file to be ingested.
+        updated_files : `set[str]`
+            Updated list of the input files with entries removed that were
+            found listed in an index file.
+        bad_index_files: `set[str]`
+            Files that looked like index files but failed to read properly.
+        """
+        # Look for index files that can speed up ingest. These can be
+        # explicit index files or index files found in directories containing
+        # files.  Group by directory and then look in those directories.
+
+        # Create set of input files to allow easy removal of entries
+        # Convert to absolute path for easy comparison with index content
+        # Do not convert to real paths since we have to assume that index
+        # files are in this location and not the location linked from
+        files_set = {os.path.abspath(f) for f in files}
+
+        # The content of all index entries indexed by full path to file
+        # to be ingested
+        index_entries = {}
+
+        # Files we failed to read
+        bad_index_files = set()
+
+        # Any good index files that were found and used
+        good_index_files = set()
+
+        # The directories containing requested files
+        by_directory = {}
+
+        for path in files_set:
+            # Keep an eye out for explicit index files
+            directory, file_in_dir = os.path.split(path)
+            if file_in_dir.endswith(".json"):
+                # Assume any explicit JSON file is an index
+                # Read the index and convert to absolute paths
+                try:
+                    index = read_index(path, force_dict=True)
+                except Exception as e:
```

Comment:
No. We explicitly want all exceptions related to reading files so that we can analyze the failures later. We don't want ingest to fall over because of one bad file if you are giving it thousands of good ones.